prov/gni/src/gnix_ep.c

```diff
@@ -120,25 +298,144 @@ struct fi_ops_tagged gnix_ep_tagged_ops;
  * EP messaging API function implementations.
  ******************************************************************************/
 
-static const recv_func_t const recv_method[] = {
-					[FI_EP_MSG] = gnix_ep_recv_msg,
-					[FI_EP_RDM] = gnix_ep_recv_rdm,
-					};
-
 static ssize_t gnix_ep_recv(struct fid_ep *ep, void *buf, size_t len,
 			    void *desc, fi_addr_t src_addr, void *context)
 {
+	int ret = FI_SUCCESS;
+	int sched_req = 0;
 	struct gnix_fid_ep *ep_priv;
+	struct gnix_fid_cq *cq;
+	struct gnix_fid_av *av;
+	struct slist_entry *item, *prev;
+	struct slist *list;
+	struct gnix_fab_req *req;
+	int matched = 0;
+	ssize_t cq_len;
+	struct gnix_address *addr_ptr;
 
 	ep_priv = container_of(ep, struct gnix_fid_ep, ep_fid);
-	assert((ep_priv->type == FI_EP_RDM) || (ep_priv->type == FI_EP_MSG));
+	if (unlikely((ep_priv->type != FI_EP_RDM) &&
+				(ep_priv->type != FI_EP_MSG)))
+		return -FI_EINVAL;
 
-	return recv_method[ep_priv->type](ep,
-					  buf,
-					  len,
-					  desc,
-					  src_addr,
-					  context);
+	if (ep_priv->type == FI_EP_RDM) {
+		av = ep_priv->av;
+		assert(av != NULL);
+		if (av->type == FI_AV_TABLE) {
+			/*
+			 * TODO: look up gni address
+			 */
+		} else
+			addr_ptr = (struct gnix_address *)&src_addr;
+	} else {
+		addr_ptr = (struct gnix_address *)&src_addr;
+	}
+
+	/*
+	 * if type FI_EP_RDM, search unexpected on this ep to see if we
+	 * already have a match. For FI_EP_MSG, just take first element
+	 * of the unexpected list.
+	 */
+
+	list = &ep_priv->unexp_recv_queue;
+	fastlock_acquire(&ep_priv->recv_queue_lock);
+
+	if (ep_priv->type == FI_EP_RDM) {
+
+		for (prev = NULL, item = list->head;
+				item; prev = item, item = item->next) {
+			req = (struct gnix_fab_req *)container_of(item,
+							  struct gnix_fab_req,
+							  slist);
+			if ((GNIX_ADDR_UNSPEC(*addr_ptr)) ||
+				(GNIX_ADDR_EQUAL(req->addr, *addr_ptr))) {
+				if (prev)
+					prev->next = item->next;
+				else
+					list->head = item->next;
+				if (!item->next)
+					list->tail = prev;
+				matched = 1;
+				req->modes |= GNIX_FAB_RQ_M_MATCHED;
+				break;
+			}
+		}
+	} else if (ep_priv->type == FI_EP_MSG) {
+
+		item = slist_remove_head(list);
+		if (item) {
+			req = (struct gnix_fab_req *)container_of(item,
+							  struct gnix_fab_req,
+							  slist);
+			req->modes |= GNIX_FAB_RQ_M_MATCHED;
+			matched = 1;
+		}
+	}
+
+	if (matched) {
+
+		if (req->modes & GNIX_FAB_RQ_M_COMPLETE) {
+			memcpy(buf, req->buf, MIN(req->len, len));
+			free(req->buf);
+			req->buf = NULL;
+		}
+
+		cq = ep_priv->recv_cq;
+		assert(cq != NULL);
```

Comment:
discussion among group and consulting man pages indicates that the application must have previously bound a  CQ to the EP.  So we'll leave the assert in.